router/internal/eventorder/eventorder.go

```diff
@@ -0,0 +1,218 @@
+package eventorder
+
+import (
+	"fmt"
+	"sync"
+)
+
+// NewBarrier creates a new properly initialized Barrier
+func NewBarrier(abortConcurrencyLimit int) *Barrier {
+	return &Barrier{
+		concurrencyLimit: abortConcurrencyLimit,
+		barriers:         make(map[string]*barrierInfo),
+	}
+}
+
+type Barrier struct {
+	concurrencyLimit int
+	mu               sync.RWMutex
+	queue            []command
+	barriers         map[string]*barrierInfo
+}
+
+// Enter the barrier for this userID and jobID. If there is not already a barrier for this userID
+// returns true, otherwise false and the previous failed jobID if this is the cause of the barrier
+func (b *Barrier) Enter(userID string, jobID int64) (accepted bool, previousFailedJobID *int64) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	barrier, ok := b.barriers[userID]
+	if !ok {
+		return true, nil
+	}
+	if barrier.failedJobID != nil {
+		failedJob := *barrier.failedJobID
+		return jobID == failedJob, &failedJob
+	}
+
+	// concurrency limiter
+	if barrier.concurrentJobs != nil {
+		barrier.mu.Lock()
+		defer barrier.mu.Unlock()
+		if _, ok := barrier.concurrentJobs[jobID]; ok {
+			return true, nil // if the job is already in the concurrent jobs map, accept it
+		}
+		if len(barrier.concurrentJobs) >= b.concurrencyLimit {
+			return false, nil // if the concurrent jobs map is full, reject it
+		}
+
+		// add the job to the concurrent jobs map and accept it
+		barrier.concurrentJobs[jobID] = struct{}{}
+		return true, nil
+	}
+
+	return true, nil
+}
+
+// Wait returns true if the job for this userID should wait (transition to a waiting state)
+func (b *Barrier) Wait(userID string, jobID int64) bool {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	barrier, ok := b.barriers[userID]
+	if !ok {
+		return false // no barrier, don't wait
+	}
+	if barrier.failedJobID != nil {
+		failedJob := *barrier.failedJobID
+		return jobID != failedJob // wait if this is not the failed job
+	}
+	// no faild job, don't wait
+	return false
+}
+
+// EnableConcurrencyLimiter enables the concurrency limiter for this barrier.
+// Should be called as soon as a job has failed in an unrecoverable manner (to be recorded as aborted)
+func (b *Barrier) EnableConcurrencyLimiter(userID string) {
```

Comment:
`[totaly minor]` Following the barrier abstraction, a simpler name _"creating a barrier for a user"_ sounds better to me.
 
```suggestion
func (b *Barrier) Create(userID string) {
```

```suggestion
func (b *Barrier) Set(userID string) {
```